src/ecdsa_impl.h

```diff
@@ -82,13 +82,13 @@ static int secp256k1_der_read_len(const unsigned char **sigp, const unsigned cha
     }
     while (lenleft > 0) {
         ret = (ret << 8) | **sigp;
-        if (ret + lenleft > (size_t)(sigend - *sigp)) {
```

Comment:
This looks like it was wrong. If the remaining digits are all zero, the final size is just `ret`, not `ret + ...`.

If my understanding is correct, is there a possibility fixing this bug could create a a consensus incompatibility?

Reply:
I don't understand your sentence about the remaining digits but I think your understanding is wrong. It took me a time to understand what this code actually does. Checking `ret > (size_t)(sigend - *sigp)` instead would be enough but checking `ret + lenleft > (size_t)(sigend - *sigp)` gives you a (not very useful) earlier return in certain cases. My understanding is that this function was correct because this function still has to read `lenleft` bytes and then the caller will have to read at least the current value of `ret`. So if the sum of these two values exceeds the remaining bytes, it is correct to return -1 early. And if we don't return -1 early, then in the last iteration we have `lenleft == 0`, and then we check with effectively only `ret` on the left-hand side, which is the correct check.

It should be noted that none of this is code is consensus-critical (for Bitcoin) because Bitcoin relies on lax DER parsing function (as implemented in the `contrib` directory of secp256k1). The change may affect other projects though. AFAIK Zcash uses the functions in the PR here.

Reply:
> This looks like it was wrong. If the remaining digits are all zero, the final size is just `ret`, not `ret + ...`.

Ah I think now I know what you are saying. It took me also a while to understand that code. The purpose of `ret  + lenleft` instead of `ret` is a (not really useful) early return: this function needs to read `lenleft` more bytes (or actually read `lenleft - 1` but we didn't increment `sigp*` yet), and the caller needs to read at least `ret` more bytes. So if `lenleft + ret` is more bytes than available, it is indeed correct to return -1 early.